File: lib/src/video_helper/exposure_helper.dart
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:flutter_webrtc/flutter_webrtc.dart';
import 'package:flutter_webrtc/src/video_helper/base/base_video_helper.dart';

class ExposureHelper extends BaseVideoHelper {
  final _exposureModeChannel = EventChannel(
    'exposureModeHandler.dataChannel',
  );
  final _exposureDurationChannel = EventChannel(
    'exposureDurationHandler.dataChannel',
  );
  final _ISOChannel = EventChannel(
    'ISOHandler.dataChannel',
  );
  final _exposureTargetBiasChannel = EventChannel(
    'exposureTargetBiasHandler.dataChannel',
  );
  final _exposureTargetOffsetChannel = EventChannel(
    'exposureTargetOffsetHandler.dataChannel',
  );

  Stream<AVCaptureExposureMode>? _exposureModeStream;
  Stream<ExposureDuration>? _exposureDurationStream;
  Stream<double>? _ISOStream;
  Stream<double>? _exposureTargetBiasStream;
  Stream<double>? _exposureTargetOffsetStream;

  Stream<AVCaptureExposureMode> get exposureModeStream {
    _exposureModeStream ??=
        _exposureModeChannel.receiveBroadcastStream().transform(
      StreamTransformer.fromHandlers(
        handleData: (index, EventSink sink) {
          sink.add(AVCaptureExposureMode.values[index]);
        },
      ),
    );
    return _exposureModeStream!;
  }

  Stream<double> get ISOStream {
    _ISOStream ??= _ISOChannel.receiveBroadcastStream().transform(
      StreamTransformer.fromHandlers(
        handleData: (data, EventSink sink) {
          sink.add(data);
        },
      ),
    );
    return _ISOStream!;
  }

  Stream<double> get exposureTargetBiasStream {
    _exposureTargetBiasStream ??=
        _exposureTargetBiasChannel.receiveBroadcastStream().transform(
      StreamTransformer.fromHandlers(
        handleData: (data, EventSink sink) {
          sink.add(data);
        },
      ),
    );
    return _exposureTargetBiasStream!;
  }

  Stream<double> get exposureTargetOffsetStream {
    _exposureTargetOffsetStream ??=
        _exposureTargetOffsetChannel.receiveBroadcastStream().transform(
      StreamTransformer.fromHandlers(
        handleData: (data, EventSink sink) {
          sink.add(data);
        },
      ),
    );
    return _exposureTargetOffsetStream!;
  }

  Stream<ExposureDuration> get exposureDurationStream {
    _exposureDurationStream ??=
        _exposureDurationChannel.receiveBroadcastStream().transform(
      StreamTransformer.fromHandlers(
        handleData: (data, EventSink sink) {
          final value = ExposureDuration.fromJson(data);
          sink.add(value);
        },
      ),
    );
    return _exposureDurationStream!;
  }

  Future<bool> isExposureModeSupported(
    AVCaptureExposureMode mode,
  ) async {
    supportedPlatforms();

    final result = await channel.invokeMethod<bool>(
      '#VideoHelper/isExposureModeSupported',
      {'mode': mode.index},
    );
    return result ?? false;
  }

  Future<bool> setExposureMode(
    AVCaptureExposureMode mode,
  ) async {
    supportedPlatforms();

    final result = await channel.invokeMethod<bool>(
      '#VideoHelper/setExposureMode',
      {'mode': mode.index},
    );
    return result ?? false;
  }

  Future<AVCaptureExposureMode> getExposureMode() async {
    supportedPlatforms();

    final index = await channel.invokeMethod<int>(
      '#VideoHelper/getExposureMode',
    );

    if (index == null || AVCaptureExposureMode.values.length < index) {
      throw 'getExposureMode error: Mode not found';
    }

    final mode = AVCaptureExposureMode.values[index];

    return mode;
  }

  Future<bool> changeISO(
    double value,
  ) async {
    supportedPlatforms();

    final result = await channel.invokeMethod<bool>(
      '#VideoHelper/changeISO',
      {'value': value},
    );
    return result ?? false;
  }

  Future<double> getISO() async {
    supportedPlatforms();

    final result = await channel.invokeMethod<double>(
      '#VideoHelper/getISO',
    );
    return result ?? 0.0;
  }

  Future<double> getExposureTargetBias() async {
    supportedPlatforms();

    final result = await channel.invokeMethod<double>(
      '#VideoHelper/getExposureTargetBias',
    );
    return result ?? 0.0;
  }

  Future<double> getMaxExposureTargetBias() async {
    supportedPlatforms();

    final result = await channel.invokeMethod<double>(
      '#VideoHelper/getMaxExposureTargetBias',
    );
    return result ?? 0.0;
  }

  Future<double> getMinExposureTargetBias() async {
    supportedPlatforms();

    final result = await channel.invokeMethod<double>(
      '#VideoHelper/getMinExposureTargetBias',
    );
    return result ?? 0.0;
  }

  Future<double> getMaxISO() async {
    supportedPlatforms();

    final result = await channel.invokeMethod<double>(
      '#VideoHelper/getMaxISO',
    );
    return result ?? 0.0;
  }

  Future<double> getExposureTargetOffset() async {
    supportedPlatforms();

    final result = await channel.invokeMethod<double>(
      '#VideoHelper/getExposureTargetOffset',
    );
    return result ?? 0.0;
  }

  Future<double> getMinISO() async {
    supportedPlatforms();

    final result = await channel.invokeMethod<double>(
      '#VideoHelper/getMinISO',
    );
    return result ?? 0.0;
  }

  Future<ExposureDuration> getExposureDuration() async {
    supportedPlatforms();

    final result = await channel.invokeMethod<Map>(
      '#VideoHelper/getExposureDuration',
    );

    if (result == null || result.length < 2) {
      throw 'getExposureDuration error: not enough parameters';
    }
    final data = ExposureDuration.fromJson(result);

    return data;
  }

  Future<ExposureDuration> getExposureDurationSeconds() async {
    supportedPlatforms();

    final result = await channel.invokeMethod<Map>(
      '#VideoHelper/getExposureDurationSeconds',
    );

    if (result == null || result.length < 2) {
      throw 'getExposureDurationSeconds error: not enough parameters';
    }
    final data = ExposureDuration.fromJson(result);

    return data;
  }

  Future<ExposureDuration> minExposureDuration() async {
    supportedPlatforms();

    final result = await channel.invokeMethod<Map>(
      '#VideoHelper/minExposureDuration',
    );

    if (result == null || result.length < 2) {
      throw 'minExposureDuration error: not enough parameters';
    }
    final data = ExposureDuration.fromJson(result);

    return data;
  }

  Future<ExposureDuration> maxExposureDuration() async {
    supportedPlatforms();

    final result = await channel.invokeMethod<Map>(
      '#VideoHelper/maxExposureDuration',
    );

    if (result == null || result.length < 2) {
      throw 'maxExposureDuration error: not enough parameters';
    }
    final data = ExposureDuration.fromJson(result);

    return data;
  }

  Future<bool> changeBias(
    double value,
  ) async {
    supportedPlatforms();

    final result = await channel.invokeMethod<bool>(
      '#VideoHelper/changeBias',
      {'value': value},
    );
    return result ?? false;
  }

  Future<bool> changeExposureDuration(
    double value,
  ) async {
    supportedPlatforms();

    final result = await channel.invokeMethod<bool>(
      '#VideoHelper/changeExposureDuration',
      {'value': value},
    );
    return result ?? false;
  }

  Future<List<AVCaptureExposureMode>> getSupportedExposureMode() async {
    supportedPlatforms();

    final result = await channel.invokeMethod<List>(
      '#VideoHelper/getSupportedExposureMode',
    );

    final supportedMode = <AVCaptureExposureMode>[];

    for (final index in result ?? []) {
      supportedMode.add(AVCaptureExposureMode.values[index]);
    }

    return supportedMode;
  }
}

File: ios/Classes/Helpers/ExposureHelper.m
//
//  ExposureHelper.m
//  flutter_webrtc
//
//  Created by Denys Dudka on 29.12.2021.
//

#import "ExposureHelper.h"
#import "FlutterDataHandler.h"
#import "FlutterWebRTCPlugin.h"

#if TARGET_OS_IPHONE
#import <Flutter/Flutter.h>
#elif TARGET_OS_MAC
#import <FlutterMacOS/FlutterMacOS.h>
#endif


@implementation ExposureHelper

__const float kExposureDurationPower = 5.0;
__const float kExposureMinimumDuration = 1.0/1000;


+(void)registerAdditionalHandlers:(NSObject<FlutterPluginRegistrar>*)registrar
                         instance:(FlutterWebRTCPlugin*)instance {
    FlutterEventChannel* exposureModeChannel = [FlutterEventChannel
                                                     eventChannelWithName:@"exposureModeHandler.dataChannel"
                                                     binaryMessenger: [registrar messenger]];
    FlutterSinkDataHandler *exposureModeHandler = [[FlutterSinkDataHandler alloc]init];
    [exposureModeChannel setStreamHandler:exposureModeHandler];
    instance.exposureModeHandler = exposureModeHandler;
    
    FlutterEventChannel* exposureDurationChannel = [FlutterEventChannel
                                                     eventChannelWithName:@"exposureDurationHandler.dataChannel"
                                                     binaryMessenger: [registrar messenger]];
    FlutterSinkDataHandler *exposureDurationHandler = [[FlutterSinkDataHandler alloc]init];
    [exposureDurationChannel setStreamHandler:exposureDurationHandler];
    instance.exposureDurationHandler = exposureDurationHandler;
    
    FlutterEventChannel* ISOChannel = [FlutterEventChannel
                                                     eventChannelWithName:@"ISOHandler.dataChannel"
                                                     binaryMessenger: [registrar messenger]];
    FlutterSinkDataHandler *ISOHandler = [[FlutterSinkDataHandler alloc]init];
    [ISOChannel setStreamHandler:ISOHandler];
    instance.ISOHandler = ISOHandler;
    
    FlutterEventChannel* exposureTargetBiasChannel = [FlutterEventChannel
                                                     eventChannelWithName:@"exposureTargetBiasHandler.dataChannel"
                                                     binaryMessenger: [registrar messenger]];
    FlutterSinkDataHandler *exposureTargetBiasHandler = [[FlutterSinkDataHandler alloc]init];
    [exposureTargetBiasChannel setStreamHandler:exposureTargetBiasHandler];
    instance.exposureTargetBiasHandler = exposureTargetBiasHandler;
    
    FlutterEventChannel* exposureTargetOffsetChannel = [FlutterEventChannel
                                                     eventChannelWithName:@"exposureTargetOffsetHandler.dataChannel"
                                                     binaryMessenger: [registrar messenger]];
    FlutterSinkDataHandler *exposureTargetOffsetHandler = [[FlutterSinkDataHandler alloc]init];
    [exposureTargetOffsetChannel setStreamHandler:exposureTargetOffsetHandler];
    instance.exposureTargetOffsetHandler = exposureTargetOffsetHandler;
}

+(void)addObservers:(AVCaptureDevice*)device
           instance:(id)instance {
    [device addObserver:instance forKeyPath:@"exposureMode" options:NSKeyValueObservingOptionNew|NSKeyValueObservingOptionOld context:nil];
    [device addObserver:instance forKeyPath:@"exposureDuration" options:NSKeyValueObservingOptionNew context:nil];
    [device addObserver:instance forKeyPath:@"ISO" options:NSKeyValueObservingOptionNew context:nil];
    [device addObserver:instance forKeyPath:@"exposureTargetBias" options:NSKeyValueObservingOptionNew context:nil];
    [device addObserver:instance forKeyPath:@"exposureTargetOffset" options:NSKeyValueObservingOptionNew context:nil];
}

+(void)removeObservers:(AVCaptureDevice*)device
              instance:(id)instance {
    [device removeObserver:instance forKeyPath:@"exposureMode" context:nil];
    [device removeObserver:instance forKeyPath:@"exposureDuration" context:nil];
    [device removeObserver:instance forKeyPath:@"ISO" context:nil];
    [device removeObserver:instance forKeyPath:@"exposureTargetBias" context:nil];
    [device removeObserver:instance forKeyPath:@"exposureTargetOffset" context:nil];
}

-(NSArray*)getSupportedExposureMode:(AVCaptureDevice*)device {
    NSMutableArray *array = [[NSMutableArray alloc] init];
    
    if ([device isExposureModeSupported:AVCaptureExposureModeLocked]) {
        [array addObject:[NSNumber numberWithInteger:AVCaptureExposureModeLocked]];
    }
    
    if ([device isExposureModeSupported:AVCaptureExposureModeAutoExpose]) {
        [array addObject:[NSNumber numberWithInteger:AVCaptureExposureModeAutoExpose]];
    }
    
    if ([device isExposureModeSupported:AVCaptureExposureModeContinuousAutoExposure]) {
        [array addObject:[NSNumber numberWithInteger:AVCaptureExposureModeContinuousAutoExposure]];
    }
    
    if ([device isExposureModeSupported:AVCaptureExposureModeCustom]) {
        [array addObject:[NSNumber numberWithInteger:AVCaptureExposureModeCustom]];
    }
    
    return array;
}

-(BOOL)setExposureMode:(AVCaptureDevice*)device
               modeNum:(NSInteger)modeNum {
    AVCaptureExposureMode mode = (AVCaptureExposureMode)modeNum;
    
    if (![self isExposureModeSupported:device modeNum: modeNum]) {
        return FALSE;
    }
    
    NSError *error;
    if ([device lockForConfiguration:&error]) {
        [device setExposureMode:mode];
        [device unlockForConfiguration];
        return TRUE;
    }
    
    if (error) {
        @throw [NSException exceptionWithName:@"Set exposure mode excetion"
                                       reason:[NSString stringWithFormat:@"%@", error]
                                     userInfo:nil];
    }
    
    return FALSE;
}

-(BOOL)isExposureModeSupported:(AVCaptureDevice*)device
                       modeNum:(NSInteger)modeNum {
    AVCaptureExposureMode mode = (AVCaptureExposureMode)modeNum;
    
    return [device isExposureModeSupported:mode];
}


-(BOOL)changeISO:(AVCaptureDevice*)device
           value:(float)value {
    NSError *error;
    
    if (value < [device activeFormat].minISO || value > [device activeFormat].maxISO) {
        NSLog(@"Change ISO excetion: value is not in minISO and maxISO range");
        return FALSE;
    }
    
    if([device lockForConfiguration:&error]) {
        [device setExposureModeCustomWithDuration:device.exposureDuration ISO:value completionHandler:nil];
        [device setWhiteBalanceMode:AVCaptureWhiteBalanceModeLocked];
        
        [device unlockForConfiguration];
        return TRUE;
    }
    if (error) {
        @throw [NSException exceptionWithName:@"Change ISO excetion"
                                       reason:[NSString stringWithFormat:@"%@", error]
                                     userInfo:nil];
    }
    return FALSE;
}

-(BOOL)changeBias:(AVCaptureDevice*)device
            value:(float)value {
    NSError *error;
    
    if (value < [device minExposureTargetBias] || value > [device maxExposureTargetBias]) {
        NSLog(@"Change bias excetion: value is not in minExposureTargetBias and maxExposureTargetBias range");
        return FALSE;
    }
    
    if([device lockForConfiguration:&error]) {
        [device setExposureTargetBias:value completionHandler:nil];
        [device setWhiteBalanceMode:AVCaptureWhiteBalanceModeLocked];
        
        [device unlockForConfiguration];
        return TRUE;
    }
    if (error) {
        @throw [NSException exceptionWithName:@"Change bias excetion"
                                       reason:[NSString stringWithFormat:@"%@", error]
                                     userInfo:nil];
    }
    return FALSE;
}

-(float)getMinISO:(AVCaptureDevice*)device {
    return [device activeFormat].minISO;
}

-(float)getMaxISO:(AVCaptureDevice*)device {
    return [device activeFormat].maxISO;
}

-(BOOL)changeExposureDuration:(AVCaptureDevice*)device
                         value:(float)value {
    
    if (device.deviceType == AVCaptureDeviceTypeBuiltInTelephotoCamera) {
        return FALSE;
    }
    
    NSError *error;
    
    if([device lockForConfiguration:&error]) {
        float iso = [self normalizeISO: device iso:device.ISO];
        float p = pow(value, kExposureDurationPower); // Apply power function to expand slider's low-end range
        float minDurationSeconds = MAX(CMTimeGetSeconds(device.activeFormat.minExposureDuration), kExposureMinimumDuration);
        float maxDurationSeconds = CMTimeGetSeconds(device.activeFormat.maxExposureDuration);
        float newDurationSeconds = p * ( maxDurationSeconds - minDurationSeconds ) + minDurationSeconds;
        
        [device setExposureModeCustomWithDuration:CMTimeMakeWithSeconds(newDurationSeconds, 1000*1000*1000) ISO:iso completionHandler:nil];
        [device setWhiteBalanceMode:AVCaptureWhiteBalanceModeLocked];
        
        [device unlockForConfiguration];
        return TRUE;
    }
    if (error) {
        @throw [NSException exceptionWithName:@"Change exposure dureation excetion"
                                       reason:[NSString stringWithFormat:@"%@", error]
                                     userInfo:nil];
    }
    return FALSE;
}

-(float)normalizeISO:(AVCaptureDevice*)device iso:(float)iso {
    float newISO = MAX(device.activeFormat.minISO, iso);
    newISO = MIN(iso, device.activeFormat.maxISO);
    
    return newISO;
}

-(CMTime)minExposureDuration:(AVCaptureDevice*)device {
    return [device activeFormat].minExposureDuration;
}

-(CMTime)maxExposureDuration:(AVCaptureDevice*)device {
    return [device activeFormat].maxExposureDuration;
}

-(AVCaptureExposureMode)getExposureMode:(AVCaptureDevice*)device {
    return [device exposureMode];
}

-(float)getExposureTargetOffset:(AVCaptureDevice*)device {
    return [device exposureTargetOffset];
}

-(float)getExposureTargetBias:(AVCaptureDevice*)device {
    return [device exposureTargetBias];
}

-(float)getMaxExposureTargetBias:(AVCaptureDevice*)device {
    return [device maxExposureTargetBias];
}

-(float)getMinExposureTargetBias:(AVCaptureDevice*)device {
    return [device minExposureTargetBias];
}


-(float)getISO:(AVCaptureDevice*)device {
    return device.ISO;
}


-(CMTime)getExposureDuration:(AVCaptureDevice*)device {
    return [device exposureDuration];
}

-(NSDictionary*)getExposureDurationSeconds:(AVCaptureDevice*)device {
    CMTime duration = [self getExposureDuration:device];
    
    return [ExposureHelper getExposureDurationSeconds:device duration:duration];
}

+(NSDictionary*)getExposureDurationSeconds:(AVCaptureDevice*)device
                                  duration:(CMTime) duration {
    CMTime min = [device activeFormat].minExposureDuration;
    CMTime max = [device activeFormat].maxExposureDuration;
    
    NSUInteger exposureDurationSeconds = CMTimeGetSeconds(duration);
    NSUInteger minExposureDurationSeconds = MAX(CMTimeGetSeconds(min), kExposureMinimumDuration);
    NSUInteger maxExposureDurationSeconds = CMTimeGetSeconds(max);
    // Map from duration to non-linear UI range 0-1
    float p = (exposureDurationSeconds - minExposureDurationSeconds) / (maxExposureDurationSeconds - minExposureDurationSeconds); // Scale to 0-1
    float value = pow(p, 1 / kExposureDurationPower);

    return @{@"duration": [NSNumber numberWithFloat:value],@"value": [NSNumber numberWithInteger:duration.value], @"timescale": [NSNumber numberWithInteger:duration.timescale], @"seconds": [NSNumber numberWithFloat:exposureDurationSeconds]};
}



@end
